### ISSUE TITLE
Hooking MSAL into OkHttp's Interceptors

### ISSUE THREAD

**Issue by 4gus71n** [AzureAD/microsoft-authentication-library-for-android#847]
**Is your feature request related to a problem? Please describe.**

I'd like to be able to fetch the MSAL auth token synchronously, so I can hook the authentication mechanism into OkHttp through an Interceptor.

Right now, on my app, all my repositories inherit from a `BaseRepository`, this `BaseRepository` has a function named `fetchMsalToken()`. This `fetchMsalToken()` function runs the `aquireTokenSilentAsync()` through RxJava. Thanks to that I can flat map this function before calling any of my endpoints, so I can be sure that I'm refreshing the auth token before each call to the backend. 

**Ideally**, I'd like to do all this within an Interceptor, somehow run the `aquireTokenSilentAsync()` synchronously, and set the auth token on the headers of the request. 

**Describe the solution you'd like**
Is it there any way to run the `aquireTokenSilentAsync()` synchronously within an Interceptor, so I can refresh the auth token there and keep any code related to this, out of my repositories.

**Describe alternatives you've considered**
As I mentioned, what I've done so far it's creating a RxJava function that runs the `aquireTokenSilentAsync()`, and using RxJava's flatmap function, I hook that into my API calls.

**Comment by iambmelt** [comment]
Hi @4gus71n -- There is a synchronous API for `acquireTokenSilent` [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java#L85).

**Comment by 4gus71n** [comment]
Thanks! I was able to move all my code into an Interceptor. 🎉 👍

**Comment by 4gus71n** [comment]
Hello! 👋 

I just recently updated to the `1.5.0` version and I started getting these errors from the MSAL library:

```
2020-06-17 14:36:18.826 20322-20322/com.example.example W/ExampleApp: com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.
2020-06-17 14:36:18.826 20322-20322/com.example.example W/ExampleApp: duplicate_command
2020-06-17 14:36:18.827 20322-20322/com.example.example W/ExampleApp: The same command was already received and is being processed.
2020-06-17 14:36:18.829 20322-20322/com.example.example W/ExampleApp: com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.
        at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:339)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:325)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:93)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:215)
        at android.app.ActivityThread.main(ActivityThread.java:6939)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:870)
```

Just to **TL;DR** what's going on here. I have an app where I have three tabs on a `ViewPager` each fragment of this `ViewPager` loads data through a different backend endpoint. I have an Interceptor on my Retrofit client, so every time I'm about to do an HTTP request I go through the MSAL library to get the authentication token.

This is the snippet of code that I use to fetch the access token:

```
    override val authToken: String
        get() {
            return publicClientApp.acquireTokenSilent(
                scope,
                publicClientApp.configuration.defaultAuthority.authorityURL.toString()
            ).accessToken
        }
```

It seems that when I do several requests in parallel this issue gets triggered.

Also, just to point out, before updating to the latest version I used to use this piece of code to fetch the authentication token:

```
    override var authToken: String?
        get() {
            return publicClientApp.acquireTokenSilent(AcquireTokenSilentParameters
                    .Builder()
                    .withScopes(scope)
                    .build()
            ).accessToken
```

But using that code trigger another error:

```
2020-06-17 14:48:00.784 22459-22459/com.example.example W/ExampleApp: current_account_mismatch
2020-06-17 14:48:00.784 22459-22459/com.example.example W/ExampleApp: The signed in account does not match with the provided account.
2020-06-17 14:48:00.787 22459-22459/com.example.example W/ExampleApp: com.microsoft.identity.client.exception.MsalClientException: The signed in account does not match with the provided account.
```

Anywho, I'm only concerned about the first error, I can "path" it from my side by keeping a reference to the access token, and using a JWT library to parse the token and know whether it's about to expire or not, but I would like to know if there's a better solution.

Thanks!

@iambmelt Please, let me know if you want me to open a new issue or if it's okay to continue this convo here. Appreciate the help! Thanks!

PS: The version that I was using before this one, was the `1.0.0`

**Comment by iambmelt** [comment]
@4gus71n The `duplicate_command` protection was added to prevent multiple identical calls simultaneously to AAD or to the local cache (which _may_ be on a remote process which can introduce latency/overhead). Basically, the library is signalling that the request you are placing has already been placed and we're still working on fetching a result

The solution to this is to place the request once, cache the result inside your app, and use it for the subsequent calls. We provide an API to check the expiry date [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msal/src/main/java/com/microsoft/identity/client/AuthenticationResult.java#L84-L101). If you know ahead of time what data you wish to display in your `ViewPager`, it may be helpful to pre-fetch the token [or tokens] you will need to render that view rather than try to load multiple, potentially duplicate tokens just-in-time

If you have more questions on `duplicate_command` I'd suggest opening a new issue and we can loop in necessary devs there

/cc @shoatman

**Comment by 4gus71n** [comment]
Thanks for the quick response @iambmelt ! 🙌 Okay, cool, I'm going to try to handle things from my side then, with that "patch". It'd be nice having the library handle these things though, the library works fine but as soon as you start trying to plug the library into some other technologies (Dagger, Retrofit) things start to work a bit funny. I get that it is equally difficult to handle these things on the library or in the client app, but maybe it would be useful introducing something like a BaseApplication class from which we could extend that guarantees that we always have an access token available. Thanks again!

**Comment by Keepertje** [comment]
> Hello! 👋
> 
> I just recently updated to the `1.5.0` version and I started getting these errors from the MSAL library:
> 
> ```
> 2020-06-17 14:36:18.826 20322-20322/com.example.example W/ExampleApp: com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.
> 2020-06-17 14:36:18.826 20322-20322/com.example.example W/ExampleApp: duplicate_command
> 2020-06-17 14:36:18.827 20322-20322/com.example.example W/ExampleApp: The same command was already received and is being processed.
> 2020-06-17 14:36:18.829 20322-20322/com.example.example W/ExampleApp: com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.
>         at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
>         at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:339)
>         at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:325)
>         at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:93)
>         at android.os.Handler.handleCallback(Handler.java:873)
>         at android.os.Handler.dispatchMessage(Handler.java:99)
>         at android.os.Looper.loop(Looper.java:215)
>         at android.app.ActivityThread.main(ActivityThread.java:6939)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:870)
> ```
> 

I am getting the same error but **only** in combination with MS Authenticator.  We are trying to make a plugin to get MSAL working with Ionic 5/ Capacitor.
